backend/src/middleware/auth.test.ts

```diff
@@ -0,0 +1,179 @@
+import jwt from "jsonwebtoken";
+import { UnauthorizedError, ExpressError } from "../expressError";
+import {
+  authenticateJWT,
+  ensureLoggedIn,
+  ensureAdmin,
+  ensureCorrectUserOrAdmin,
+} from "./auth";
+import { Request, Response, NextFunction } from "express";
+
+import { SECRET_KEY } from "../config";
+
+const testJwt = jwt.sign({ username: "test", isAdmin: false }, SECRET_KEY);
+const badJwt = jwt.sign({ username: "test", isAdmin: false }, "wrong");
+
+describe("authenticateJWT", function () {
+  test("works: via header", function () {
+    expect.assertions(2);
+    const req = {
+      headers: { authorization: `Bearer ${testJwt}` },
+    };
+    const res = { locals: {} };
+    const next = function (err: ExpressError) {
+      expect(err).toBeFalsy();
+    };
+    authenticateJWT(req as Request, res as Response, next as NextFunction);
+    expect(res.locals).toEqual({
+      user: {
+        iat: expect.any(Number),
+        username: "test",
+        isAdmin: false,
+      },
+    });
+  });
+
+  test("works: no header", function () {
+    expect.assertions(2);
+    const req = {};
+    const res = { locals: {} };
+    const next = function (err: ExpressError) {
+      expect(err).toBeFalsy();
+    };
+    authenticateJWT(req as Request, res as Response, next as NextFunction);
+    expect(res.locals).toEqual({});
+  });
+
+  test("works: invalid token", function () {
+    expect.assertions(2);
+    const req = { headers: { authorization: `Bearer ${badJwt}` } };
+    const res = { locals: {} };
+    const next = function (err: ExpressError) {
+      expect(err).toBeFalsy();
+    };
+    authenticateJWT(req as Request, res as Response, next as NextFunction);
+    expect(res.locals).toEqual({});
+  });
+});
+
+describe("ensureLoggedIn", function () {
+  test("works", function () {
+    expect.assertions(1);
+    const req = {};
+    const res: unknown = { locals: { user: { username: "test" } } };
```

Comment:
Utilizing `unknown` for various `req` and `res` when type casting as `Request` and `Response` weren't enough. Need to look this up to see if it's best practice. 

![TS Error](https://user-images.githubusercontent.com/69769431/227858275-def64534-b41d-41cb-9ed1-2635fe926a83.png)